sdks/python/apache_beam/transforms/core.py

```diff
@@ -119,6 +120,112 @@ def aggregate_to(self, aggregator, input_value):
     self.state.counter_for(aggregator).update(input_value)
 
 
+class NewDoFn(WithTypeHints, HasDisplayData):
+  """A function object used by a transform with custom processing.
+
+  The ParDo transform is such a transform. The ParDo.apply
+  method will take an object of type DoFn and apply it to all elements of a
+  PCollection object.
+
+  In order to have concrete DoFn objects one has to subclass from DoFn and
+  define the desired behavior (start_bundle/finish_bundle and process) or wrap a
+  callable object using the CallableWrapperDoFn class.
+  """
+
+  ElementParam = 'ElementParam'
+  ContextParam = 'ContextParam'
+  SideInputParam = 'SideInputParam'
+  TimestampParam = 'TimestampParam'
+  WindowsParam = 'WindowsParam'
```

Comment:
For the record, the fact that multiple windows are optimized in this way should be transparent to the user. If the window is exposed, or a side input depends on the window, we must call process multiple times.